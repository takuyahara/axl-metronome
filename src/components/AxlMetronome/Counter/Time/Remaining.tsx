import React, { Component, createRef, RefObject } from 'react';
import _ from 'lodash';
import style from './Remaining.module.scss';
import Utils, { ICssSelector } from 'utils/Utils';

interface IProps {
  timeToCount: number;
  postProcess?: IPostProcess;
  inheritedSelector?: ICssSelector;
}
interface IState {
  remaining: number;
}
interface IPostProcess {
  onClick?: (e: React.MouseEvent) => void;
  onTouchEnd?: (e: React.TouchEvent) => void;
}
class Remaining extends Component<IProps, IState> {
  private postClick: (e: React.MouseEvent) => void;
  private postTouchEnd: (e: React.TouchEvent) => void;

  public constructor(props: IProps) {
    super(props);
    this.state = {
      remaining: props.timeToCount,
    };
    this.postClick = props.postProcess && props.postProcess.onClick ? props.postProcess.onClick : (e: React.MouseEvent) => {};
    this.postTouchEnd = props.postProcess && props.postProcess.onTouchEnd ? props.postProcess.onTouchEnd : (e: React.TouchEvent) => {};

    this.onClick = this.onClick.bind(this);
    this.onTouchEnd = this.onTouchEnd.bind(this);
  }
  private onClick(e: React.MouseEvent): void {
    this.postClick(e);
  }
  private onTouchEnd(e: React.TouchEvent): void {
    this.postTouchEnd(e);
  }
  private formatTime(timeInSecond: number): string
  {
    const minute = Math.floor(timeInSecond / 60);
    const second = Math.floor(timeInSecond % 60);
    const minuteStr = minute < 10 ? "0" + minute : "" + minute;
    const secondStr = second < 10 ? "0" + second : "" + second;
    return minuteStr + ":" + secondStr;
  }
  public setRemaining(remaining: number): void {
    this.setState({
      remaining,
    });
  }
  /* istanbul ignore next */
  public init(newProps: IProps): void {
    this.setState({
      remaining: newProps.timeToCount,
    });
  }
  /* istanbul ignore next */
  public componentWillMount(): void {
    if (this.props.inheritedSelector) {
      Utils.inheritSelector(style, this.props.inheritedSelector);
    }
  }
  /* istanbul ignore next */
  public componentWillReceiveProps(newProps: IProps): void {
    if (_.isEqual(this.props, newProps)) {
      return;
    }
    this.init(newProps);
  }
  public render(): React.ReactNode {
    const formattedTime = this.formatTime(Math.round(this.state.remaining));
    return (
      <div 
        className={style.indicator} 
        data-testid="indicator-remaining"
        onClick={this.onClick}
        onTouchEnd={this.onTouchEnd}
      >
        <div className={style.remaining}>
          {formattedTime}
        </div>
      </div>
    );
  }
}
export default Remaining;
