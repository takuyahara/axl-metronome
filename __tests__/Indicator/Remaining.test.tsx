import React from 'react';
import { configure, mount, ReactWrapper } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import sinon from 'sinon';
import Remaining from 'Indicator/Remaining';

interface ISpy {
  [s: string]: sinon.SinonSpy;
}

configure({ adapter: new Adapter() });

// tslint:disable:no-any
describe('Indicator', () => {
  let remaining: ReactWrapper;
  let remainingInstance: Remaining;
  let spy: ISpy = {};
  
  afterEach(() => {
    remaining.update();
    expect(remaining).toMatchSnapshot();
    sinon.restore();
    remaining.unmount();
  });

  describe('without post process', () => {
    beforeEach(() => {
      remaining = mount(<Remaining 
        timeToCount={900} 
      />);
      remainingInstance = remaining.instance() as Remaining;
      sinon.resetHistory();
    });

    it('formats number', () => {
      expect((remainingInstance as any).formatTime(0)).toBe('00:00');
      expect((remainingInstance as any).formatTime(9)).toBe('00:09');
      expect((remainingInstance as any).formatTime(10)).toBe('00:10');
      expect((remainingInstance as any).formatTime(59)).toBe('00:59');
      expect((remainingInstance as any).formatTime(60)).toBe('01:00');
      expect((remainingInstance as any).formatTime(69)).toBe('01:09');
      expect((remainingInstance as any).formatTime(70)).toBe('01:10');
      expect((remainingInstance as any).formatTime(600)).toBe('10:00');
      expect((remainingInstance as any).formatTime(3599)).toBe('59:59');
    });
  
    it('setRemaining', () => {
      remainingInstance.setRemaining(60);
    });

    describe('MouseEvent', () => {
      it('click', () => {
        const postClick = sinon.spy(remainingInstance as any, 'postClick');
        remaining.find('.remaining').simulate('click');
        expect(postClick.calledOnce).toBe(true);
      });  
    });

    describe('TouchEvent', () => {
      it('touchEnd', () => {
        const postTouchEnd = sinon.spy(remainingInstance as any, 'postTouchEnd');
        remaining.find('.remaining').simulate('touchend');
        expect(postTouchEnd.calledOnce).toBe(true);
      });  
    });
  });

  describe('with post process', () => {
    beforeEach(() => {
      spy = {
        onClick: sinon.spy(),
        onTouchEnd: sinon.spy(),
      };
      remaining = mount(<Remaining 
        timeToCount={900} 
        postProcess={spy} 
      />);
      remainingInstance = remaining.instance() as Remaining;
      sinon.resetHistory();
    });

    describe('MouseEvent', () => {
      it('invokes postClick', () => {
        remaining.find('.remaining').simulate('click');
        expect(spy.onClick.calledOnce).toBe(true);
      });
    });

    describe('TouchEvent', () => {
      it('invokes postTouchEnd', () => {
        remaining.find('.remaining').simulate('touchend');
        expect(spy.onTouchEnd.calledOnce).toBe(true);
      });
    });
  });
});
