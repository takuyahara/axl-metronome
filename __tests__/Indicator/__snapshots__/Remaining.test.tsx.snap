// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indicator with post process MouseEvent invokes postClick 1`] = `
<Remaining
  postProcess={
    Object {
      "onClick": [Function],
      "onTouchEnd": [Function],
    }
  }
  timeToCount={900}
>
  <div
    className="indicator"
    onClick={[Function]}
    onTouchEnd={[Function]}
  >
    <div
      className="remaining"
    >
      15:00
    </div>
  </div>
</Remaining>
`;

exports[`Indicator with post process TouchEvent invokes postTouchEnd 1`] = `
<Remaining
  postProcess={
    Object {
      "onClick": [Function],
      "onTouchEnd": [Function],
    }
  }
  timeToCount={900}
>
  <div
    className="indicator"
    onClick={[Function]}
    onTouchEnd={[Function]}
  >
    <div
      className="remaining"
    >
      15:00
    </div>
  </div>
</Remaining>
`;

exports[`Indicator without post process MouseEvent click 1`] = `
<Remaining
  timeToCount={900}
>
  <div
    className="indicator"
    onClick={[Function]}
    onTouchEnd={[Function]}
  >
    <div
      className="remaining"
    >
      15:00
    </div>
  </div>
</Remaining>
`;

exports[`Indicator without post process TouchEvent touchEnd 1`] = `
<Remaining
  timeToCount={900}
>
  <div
    className="indicator"
    onClick={[Function]}
    onTouchEnd={[Function]}
  >
    <div
      className="remaining"
    >
      15:00
    </div>
  </div>
</Remaining>
`;

exports[`Indicator without post process formats number 1`] = `
<Remaining
  timeToCount={900}
>
  <div
    className="indicator"
    onClick={[Function]}
    onTouchEnd={[Function]}
  >
    <div
      className="remaining"
    >
      15:00
    </div>
  </div>
</Remaining>
`;

exports[`Indicator without post process setRemaining 1`] = `
<Remaining
  timeToCount={900}
>
  <div
    className="indicator"
    onClick={[Function]}
    onTouchEnd={[Function]}
  >
    <div
      className="remaining"
    >
      01:00
    </div>
  </div>
</Remaining>
`;
